src/main/java/be/seeseemelk/mockbukkit/entity/AllayMock.java

```diff
@@ -150,78 +154,79 @@ public void assertInventoryContains(ItemStack item, String s)
 	@Override
 	public boolean canDuplicate()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.canDuplicate;
 	}
 
 	@Override
 	public void setCanDuplicate(boolean canDuplicate)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		this.canDuplicate = canDuplicate;
 	}
 
 	@Override
 	public long getDuplicationCooldown()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.duplicationCoolDown;
 	}
 
 	@Override
 	public void setDuplicationCooldown(long cooldown)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		this.duplicationCoolDown = cooldown;
 	}
 
 	@Override
 	public void resetDuplicationCooldown()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		this.duplicationCoolDown = 6000L;
+		this.canDuplicate = false;
 	}
 
 	@Override
 	public boolean isDancing()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.isDancing;
 	}
 
 	@Override
 	public void startDancing(@NotNull Location location)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkArgument(location != null, "Location cannot be null");
+		Preconditions.checkArgument(location.getBlock().getType() == Material.JUKEBOX,
+				"Location must be a Jukebox");
+
+		this.isDancing = true;
+		this.jukebox = location.clone();
 	}
 
 	@Override
 	public void startDancing()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		this.isDancing = true;
 	}
 
 	@Override
 	public void stopDancing()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		this.isDancing = false;
+		this.jukebox = null;
 	}
 
 	@Override
 	public @Nullable Allay duplicateAllay()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Allay allay = (Allay) this.getWorld().spawn(this.getLocation(), Allay.class,null,
+				CreatureSpawnEvent.SpawnReason.DUPLICATION);
+		allay.resetDuplicationCooldown();
+		this.resetDuplicationCooldown();
+
+		return allay;
 	}
 
 	@Override
 	public @Nullable Location getJukebox()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.jukebox;
```

Comment:
```suggestion
		return this.jukebox.clone();
```
You should also do this (same reasoning as in startdancing)